### ISSUE TITLE
filter2manager single host icmp bug

### ISSUE THREAD

**Issue by Metallion** [axsh/openvnet#431]
### TL/DR

when pinging from an interface without filter enabled to one with filtering enabled, the ping replies are blocked at the destination, despite filtering being disabled. This occurs on interfaces connected to the same datapath.

### Problem

I tried out filter2manager with the installation guide environment. That gives me `inst1` and `inst2`in one host on the same subnet.

I enabled filtering on `inst1` and opened both icmp and tcp for all source IPs.

```
[18:36:18] centos@joske  (/opt/axsh/openvnet/client/vnctl) > vnctl filters show
---
total_count: 1
offset: 0
limit: 30
:items:
- :id: 1
  :uuid: fil-test1
  :mode: static
  :interface_id: 1
  :ingress_passthrough: false
  :egress_passthrough: false
  :created_at: 2015-12-21 08:53:24.000000000 Z
  :updated_at: 2015-12-21 08:53:24.000000000 Z
  :deleted_at: 
  :is_deleted: 0
  ! ':': 
  :interface_uuid: if-inst1


[18:43:17] centos@joske  (/opt/axsh/openvnet/client/vnctl) > vnctl filters static show
---
total_count: 2
offset: 0
limit: 30
:items:
- :id: 1
  :filter_id: 1
  :ipv4_src_address: 0.0.0.0
  :ipv4_dst_address: 0.0.0.0
  :ipv4_src_prefix: 0
  :ipv4_dst_prefix: 0
  :port_src: 0
  :port_dst: 0
  :protocol: tcp
  :passthrough: true
  :created_at: 2015-12-21 08:55:53.000000000 Z
  :updated_at: 2015-12-21 08:55:53.000000000 Z
  :deleted_at: 
  :is_deleted: 0
  ! ':': 
- :id: 2
  :filter_id: 1
  :ipv4_src_address: 0.0.0.0
  :ipv4_dst_address: 0.0.0.0
  :ipv4_src_prefix: 0
  :ipv4_dst_prefix: 0
  :port_src: 
  :port_dst: 
  :protocol: icmp
  :passthrough: true
  :created_at: 2015-12-21 08:57:50.000000000 Z
  :updated_at: 2015-12-21 08:57:50.000000000 Z
  :deleted_at: 
  :is_deleted: 0
  ! ':':
```

ssh from `inst2` to `inst1` works.

When pinging from `inst2` to `inst2`, the ping reply is blocked when returning to `inst2`. tcpdump shows that ping arrives at `inst1`.

```
[18:43:31] centos@joske  (/opt/axsh/openvnet/client/vnctl) > sudo tcpdump -i inst1
[sudo] password for centos: 
tcpdump: WARNING: inst1: no IPv4 address assigned
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on inst1, link-type EN10MB (Ethernet), capture size 65535 bytes
18:47:07.428092 ARP, Request who-has 10.100.0.10 tell 10.100.0.11, length 28
18:47:07.428119 ARP, Reply 10.100.0.10 is-at 10:54:ff:00:00:01 (oui Unknown), length 28
18:47:07.430245 IP 10.100.0.11 > 10.100.0.10: ICMP echo request, id 64256, seq 1, length 64
18:47:07.430285 IP 10.100.0.10 > 10.100.0.11: ICMP echo reply, id 64256, seq 1, length 64
```

vnflows-monitor shows that the reply is being dropped in the INGRESS_FILTER table.

```
-------run:22--iteration:2-------
(0): TABLE_CLASSIFIER
  0-02      113       0 => PORT(0x1)                in_port=1 actions=write_metadata:TYPE_INTERFACE|LOCAL(0x1),goto_table:TABLE_INTERFACE_EGRESS_CLASSIFIER(15)
  0-02       79       0 => PORT(0x2)                in_port=2 actions=write_metadata:TYPE_INTERFACE|LOCAL(0x2),goto_table:TABLE_INTERFACE_EGRESS_CLASSIFIER(15)
(15): TABLE_INTERFACE_EGRESS_CLASSIFIER
 15-30      112       0 => INTERFACE(0x1)[0x12]     metadata=TYPE_INTERFACE(0x1) actions=goto_table:TABLE_INTERFACE_EGRESS_VALIDATE(17)
 15-30       79       0 => INTERFACE(0x2)[0x12]     metadata=TYPE_INTERFACE(0x2) actions=goto_table:TABLE_INTERFACE_EGRESS_VALIDATE(17)
(17): TABLE_INTERFACE_EGRESS_VALIDATE
 17-30      106       0 => INTERFACE(0x1)[0x12]     ip,metadata=TYPE_INTERFACE(0x1),dl_src=10:54:ff:00:00:01,nw_src=10.100.0.10 actions=write_metadata:TYPE_NETWORK(0x1),goto_table:TABLE_NETWORK_CONNECTION(20)
 17-30       73       0 => INTERFACE(0x2)[0x12]     ip,metadata=TYPE_INTERFACE(0x2),dl_src=10:54:ff:00:00:02,nw_src=10.100.0.11 actions=write_metadata:TYPE_NETWORK(0x1),goto_table:TABLE_NETWORK_CONNECTION(20)
(20): TABLE_NETWORK_CONNECTION
 20-00      191       0 => SWITCH(0x0)               actions=goto_table:TABLE_NETWORK_SRC_CLASSIFIER(21)
(21): TABLE_NETWORK_SRC_CLASSIFIER
 21-30      191       0 => NETWORK(0x1)             metadata=TYPE_NETWORK(0x1) actions=goto_table:TABLE_ROUTE_INGRESS_INTERFACE(30)
(30): TABLE_ROUTE_INGRESS_INTERFACE
 30-00      191       0 => SWITCH(0x0)               actions=goto_table:TABLE_NETWORK_DST_CLASSIFIER(42)
(42): TABLE_NETWORK_DST_CLASSIFIER
 42-30      191       0 => NETWORK(0x1)             metadata=TYPE_NETWORK(0x1) actions=goto_table:TABLE_NETWORK_DST_MAC_LOOKUP(43)
(43): TABLE_NETWORK_DST_MAC_LOOKUP
 43-60       79       0 => INTERFACE(0x1)[0x12]     metadata=TYPE_NETWORK(0x1),dl_dst=10:54:ff:00:00:01 actions=write_metadata:TYPE_INTERFACE(0x1),goto_table:TABLE_INTERFACE_INGRESS_FILTER(45)
 43-60      111       0 => INTERFACE(0x2)[0x12]     metadata=TYPE_NETWORK(0x1),dl_dst=10:54:ff:00:00:02 actions=write_metadata:TYPE_INTERFACE(0x2),goto_table:TABLE_INTERFACE_INGRESS_FILTER(45)
(45): TABLE_INTERFACE_INGRESS_FILTER
 45-00      106       0 => SWITCH(0x0)               actions=goto_table:TABLE_INTERFACE_INGRESS_FILTER_LOOKUP(46)
 45-90       73       0 => INTERFACE(0x1)           metadata=TYPE_INTERFACE(0x1) actions=goto_table:TABLE_OUT_PORT_INTERFACE_INGRESS(90)
(46): TABLE_INTERFACE_INGRESS_FILTER_LOOKUP
 46-00      106       0 => SWITCH(0x0)               actions=drop
(90): TABLE_OUT_PORT_INTERFACE_INGRESS
 90-10       79       0 => PORT(0x1)                metadata=TYPE_INTERFACE(0x1) actions=output:1
```

This is the contents of the ingress filter table.

```
 43-60      289       0 => INTERFACE(0x1)[0x12]     metadata=TYPE_NETWORK(0x1),dl_dst=10:54:ff:00:00:01 actions=write_metadata:TYPE_INTERFACE(0x1),goto_table:TABLE_INTERFACE_INGRESS_FILTER(45)
 43-60      321       0 => INTERFACE(0x2)[0x12]     metadata=TYPE_NETWORK(0x1),dl_dst=10:54:ff:00:00:02 actions=write_metadata:TYPE_INTERFACE(0x2),goto_table:TABLE_INTERFACE_INGRESS_FILTER(45)
(45): TABLE_INTERFACE_INGRESS_FILTER
 45-00      304       0 => SWITCH(0x0)               actions=goto_table:TABLE_INTERFACE_INGRESS_FILTER_LOOKUP(46)
 45-10        0       0 => FILTER2(0x1)             metadata=TYPE_INTERFACE(0x1) actions=drop
 45-21        0       0 => FILTER2(0x1)             icmp,metadata=TYPE_INTERFACE(0x1) actions=goto_table:TABLE_OUT_PORT_INTERFACE_INGRESS(90)
 45-21        0       0 => FILTER2(0x1)             tcp,metadata=TYPE_INTERFACE(0x1) actions=goto_table:TABLE_OUT_PORT_INTERFACE_INGRESS(90)
 45-90      271       0 => INTERFACE(0x1)           metadata=TYPE_INTERFACE(0x1) actions=goto_table:TABLE_OUT_PORT_INTERFACE_INGRESS(90)
 45-90       35       0 => FILTER(0x0)[0x11]        arp actions=goto_table:TABLE_OUT_PORT_INTERFACE_INGRESS(90)
(46): TABLE_INTERFACE_INGRESS_FILTER_LOOKUP
 46-00      304       0 => SWITCH(0x0)               actions=drop
(50): TABLE_FLOOD_SIMULATED
 50-00        1       0 => SWITCH(0x0)               actions=goto_table:TABLE_FLOOD_LOCAL(51)
(51): TABLE_FLOOD_LOCAL
 51-00        0       0 => SWITCH(0x0)               actions=drop
```

### Solution

* First write an integration test that reproduces the problem. Currently the integration test does not have two interfaces on the same datapath with one that has filtering enabled and one that hasn't.

* Debug and fix it.

**Comment by toros11** [comment]
Reproducing the conditions as an integration test it got resolved by adding an arp rule.

The old filter manager only adds one by default into the INGRESS_FILTER table, which would explain why the packages are received as but not delivered.
